### PR TITLE
[feature] Support assertion if conditional is IPv4/IPv6 only and partial perusal of block entries

### LINKED ISSUE #48: Support sub-addressing of GPFile parts for time range and IPv4/IPv6 conditions
There's two scenarios where it could be useful to support reading / addressing only parts of the source files:

- If there's a time range overlap with the current directory (i.e. "day") that would cause only parts of the file being used (in particular when using the memory buffered file) as mentioned by @els0r .
- If there's a conditional that is either IPv4-only or IPv6-only. Since the flows are split at a margin it would significantly boost performance for such cases if GPFile would only consume / read the required part of the files and the `DBWorkManager` would start / stop only at the required ranges.

DoD:

- [ ] Determine if a query has conditionals that limit it to IPv4 / IPv6 only
- [ ] Support sub-addressing the MemFile (i.e. handing over a start / end option to GPFile that limits either the os.File or the MemFile to the required range, causing only a partial read)
- [ ] Set the loop boundaries and read operations in `DBWorkManager` accordingly

### PR DESCRIPTION
**Pull request by fako1024**
So, this is currently only a draft because I hit a slight bump. The condition check was actually quite nicely doable within the context of the conditional `Node`. Same goes for setting loop limits inside the individual work to be done inside blocks.

But: With the current setup I cannot limit disk access / decompression to IPv4 / IPv6 only (simply because both IPv4 and IPv6 data is encoded & compressed together in a single block operation), in order to support that I'd have to adapt the DB format.

Options I see:

- Internally split the data blocks into two (IPv4 / IPv6) and compress / make them available individually. Would be mostly under the hood of `GPDir` / `GPFile` and slightly increase the metadata size (because I'd have to keep track of two compression blocks per actual block) and add some rather nasty `seek()` logic.
- Store IPv4 / IPv6 data in separate files (either another subdirectory or next to each other, won't really make a difference). Would be mostly type work (because we're basically doubling the number of files, but apart from that it's probably going to be some extensions to the conditional logic on what to read and what not). Would double the amount of file / disk IOPS for non-conditional / dual-stack queries. Advantages would be better compression ratio (because IPv4 / IPv6 would be in their own "entropy pools", so lz4 / zstd could build better dictionaries) and little additional complication (because I only need to decide which files to open, not do some seek magic).
- Do nothing and take the little improvement from this existing PR changes (always read everything, just don't loop over the input). In practice it's a rather small improvement (most prominent of course when selection IPv6 only) since the current logic would _loop_ over the excluded entries but pretty quickly call `continue` because the condition isn't met (so we're basically saving some loop counter iterations and conditional evaluations including their `memcpy()` operations from the block to the comparison key variable).

@els0r What's your opinion on this? I don't have a good feeling for practical uses, but I guess filtering for an IP is going to be one of the most common use cases. So especially when looking for few IPs in a large set of data being able to limit data accesses would be a nice thing (especially if in the future IPv6 becomes more prevalent, should that ever happen)...

Closes #48

### CODE REVIEW

**Comment by els0r** [comment]
Hey. Thanks for looking into this. My take:

> Options I see:
> 
> * Internally split the data blocks into two (IPv4 / IPv6) and compress / make them available individually. Would be mostly under the hood of `GPDir` / `GPFile` and slightly increase the metadata size (because I'd have to keep track of two compression blocks per actual block) and add some rather nasty `seek()` logic.

Nasty `seek()` logic doesn't sound enticing and will only make the code less readable. Not worth it, IMHO.

> * Store IPv4 / IPv6 data in separate files (either another subdirectory or next to each other, won't really make a difference). Would be mostly type work (because we're basically doubling the number of files, but apart from that it's probably going to be some extensions to the conditional logic on what to read and what not). Would double the amount of file / disk IOPS for non-conditional / dual-stack queries. Advantages would be better compression ratio (because IPv4 / IPv6 would be in their own "entropy pools", so lz4 / zstd could build better dictionaries) and little additional complication (because I only need to decide which files to open, not do some seek magic).

If I understand correctly, we'd have 4 files: `{s, d}ip_ipv{4, 6}.gpf`, right? If the additional IOPS for the non-conditional queries aren't impacting query performance, I'd say we go for it. Because of further disk size improvements. And because it would speed up the conditional queries. In almost 10 years, I have yet to see a query being made for IPv4 and IPv6 at the same time, so there's something to gain.

> * Do nothing and take the little improvement from this existing PR changes (always read everything, just don't loop over the input). In practice it's a rather small improvement (most prominent of course when selection IPv6 only) since the current logic would _loop_ over the excluded entries but pretty quickly call `continue` because the condition isn't met (so we're basically saving some loop counter iterations and conditional evaluations including their `memcpy()` operations from the block to the comparison key variable).

If time becomes and issue we can also do this.

I strongly prefer option 2, evaluate performance and then release it with v4 if viable. Just a question if this is a big task or rather contained. What do you think?

**Comment by els0r** [comment]
Decision made not to engage further in the scope of v4.

**Comment by fako1024** [comment]
@els0r Alright, rebased / merged develop and fixed conflicts, ready to review....

File: pkg/types/types.go [hunk 1]
```diff
@@ -86,15 +119,15 @@ func numZeros(ip []byte) uint8 {
 }
 
 // IPStringToBytes creates a goDB compatible bytes slice from an IP address string
-func IPStringToBytes(ip string) ([]byte, error) {
+func IPStringToBytes(ip string) ([]byte, bool, error) {
```
Comment (els0r) [inline pkg/types/types.go:122]: Nitpick: this function could benefit from either named return parameters or a note in the description that we return whether the IP string is an IPv4 string.

**Review by els0r — APPROVED** [review]
Me gusta!

File: pkg/types/types.go (hunk 1 above)
Comment (fako1024) [reply pkg/types/types.go]: Fair point. Did both and added a few simple unit tests.

**Pull request merged**